mdoc/Mono.Documentation/MDocUpdater.cs

```diff
@@ -2518,25 +2518,80 @@ private void UpdateMember (DocsNodeInfo info, FrameworkTypeEntry typeEntry, Dict
             UpdateExtensionMethods (me, info);
         }
 
-        private static void UpdateSignature(MemberFormatter formatter, TypeDefinition type, XmlElement xmlElement)
-        {
-            var usageSample = formatter.UsageFormatter?.GetDeclaration(type);
-            var valueToUse = formatter.GetDeclaration (type);
-            if (valueToUse == null && usageSample == null)
-                return;
-
-            string elementXPath = "TypeSignature[@Language='" + formatter.Language + "']";
-            GetUpdateXmlMethods(formatter, xmlElement, elementXPath, valueToUse, usageSample,
-                out var valueMatches,
-                out var setValue,
-                out var makeNewNode);
-
-            AddXmlNode(
-                xmlElement.SelectNodes (elementXPath).Cast<XmlElement> ().ToArray (),
-                valueMatches,
-                setValue,
-                () => makeNewNode(true),
-                type);
+        private static void UpdateSignature(MemberFormatter formatter, TypeDefinition type, XmlElement xmlElement, FrameworkTypeEntry typeEntry)
+        {
+            string elementName = "TypeSignature";
+            string elementXPath = $"{elementName}[@Language='" + formatter.Language + "']";
+            var existingElements = QueryXmlElementsByXpath(xmlElement, elementXPath).ToList();
+
+            if (typeEntry.TimesProcessed > 1 && existingElements.Any())
+                return;
+
+            // if first framework of type, clear signatures and generate from scratch
+            if (typeEntry.Framework.IsFirstFrameworkForType(typeEntry))
+            {
+                existingElements.ForEach(e => e.ParentNode.RemoveChild(e));
+                existingElements = QueryXmlElementsByXpath(xmlElement, elementXPath).ToList();
+            }
+
+            // start to process 
+            var usageSample = formatter.UsageFormatter?.GetDeclaration(type);
+            var valueToUse = formatter.GetDeclaration(type);
+            if (valueToUse == null && usageSample == null)
+                return;
+
+            bool elementFound = false;
+
+            // if there is already signature found under same formatter, update fxa
+            if (existingElements.Any())
+            {
+                foreach (var element in existingElements)

```

Comment:
ok, so I'm comparing this to the similar `MemberSignature` method, and it looks like this particular section isn't complete. This is what _that_ section looks like:

```csharp
                foreach(var element in existingElements)
                {
                    var val = element.GetAttribute("Value");
                    var usage = element.GetAttribute("Usage");
                    if (val == (valueToUse??"") && usage == (usageSample??""))
                    {
                        // add FXA
                        string newfxa = FXUtils.AddFXToList(element.GetAttribute(Consts.FrameworkAlternate), typeEntry.Framework.Name);
                        element.SetAttribute(Consts.FrameworkAlternate, newfxa);
                        elementFound = true;
                    }
                    else
                    {
                        string newfxa = FXUtils.RemoveFXFromList(element.GetAttribute(Consts.FrameworkAlternate), typeEntry.Framework.Name);
                        element.SetAttribute(Consts.FrameworkAlternate, newfxa);
                    }
                }
```

Note the _else_ branch, which will remove the framework from that element (if it's there) if `val` and `usage` don't match.

Actually ... now that I'm working through this, I wonder if this is just an artifact of the previous iteration where it wasn't clearing the list every time. Is this `else` in the other method necessary?

Reply:
as discussed ... this `else` is not necessary